Скрит позволяет разбивать изображения на части для обучения

In [14]:
import os
import cv2
import numpy as np
from tqdm import tqdm
import random

# Параметры разбиения
IMAGE_DIR = "../DataSet/SplitDataAll/Valid/images/"
LABEL_DIR = "../DataSet/SplitDataAll/Valid/labels/"

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".bmp", ".tiff"}

In [17]:
def count_images_without_labels(image_dir, label_dir):
    count = 0

    for file in os.listdir(image_dir):
        if os.path.splitext(file)[1].lower() in IMAGE_EXTENSIONS:
            image_name = os.path.splitext(file)[0]
            label_path = os.path.join(label_dir, f"{image_name}.txt")

            # Проверяем, существует ли файл метки и не пустой ли он
            if not os.path.exists(label_path) or os.path.getsize(label_path) == 0:
                count += 1

    return count

missing_labels = count_images_without_labels(IMAGE_DIR, LABEL_DIR)
print(f"Количество изображений без меток: {missing_labels}")


Количество изображений без меток: 54624


train - Количество изображений без меток: 477548 было и 166877 стало из 172475
valid - Количество изображений без меток: 136363 было и 54624 стало из 56169
test  - Количество изображений без меток: 68204 было из 69000 и 34418 стало из 35214

In [16]:

def delete_images_without_labels(image_dir, label_dir, probability=0.5):
    """
    Удаляет изображения, у которых нет меток, с заданной вероятностью.

    :param image_dir: Папка с изображениями.
    :param label_dir: Папка с метками (если метки в той же папке, указать image_dir).
    :param probability: Вероятность удаления (от 0 до 1).
    """
    IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".bmp", ".tiff"}
    deleted_count = 0

    for file in os.listdir(image_dir):
        if os.path.splitext(file)[1].lower() in IMAGE_EXTENSIONS:
            image_name = os.path.splitext(file)[0]
            label_path = os.path.join(label_dir, f"{image_name}.txt")
            image_path = os.path.join(image_dir, file)

            # Если файла метки нет или он пустой
            if not os.path.exists(label_path) or os.path.getsize(label_path) == 0:
                if random.random() < probability:  # Удаляем с вероятностью probability
                    os.remove(image_path)
                    deleted_count += 1

    print(f"Удалено {deleted_count} изображений без меток.")

delete_images_without_labels(IMAGE_DIR, LABEL_DIR, probability=0.6)

Удалено 81739 изображений без меток.
